src/stream/src/executor/source/fetch_executor.rs

```diff
@@ -0,0 +1,367 @@
+        );
+
+        if barrier.is_pause_on_startup() {
+            stream.pause_stream();
+        }
+
+        // If it is a recovery startup,
+        // there can be file assignments in the state table.
+        // Hence we try building a reader first.
+        Self::try_replace_with_new_batch_reader(
+            &mut splits_on_fetch,
+            &state_store_handler,
+            core.column_ids.clone(),
+            self.build_source_ctx(&source_desc, core.source_id),
+            &source_desc,
+            &mut stream,
+        )
+        .await?;
+
+        yield Message::Barrier(barrier);
+
+        while let Some(msg) = stream.next().await {
+            match msg {
+                Err(e) => {
+                    tracing::error!("Fetch Error: {:?}", e);
+                    splits_on_fetch = 0;
+                    Self::try_replace_with_new_batch_reader(
+                        &mut splits_on_fetch,
+                        &state_store_handler,
+                        core.column_ids.clone(),
+                        self.build_source_ctx(&source_desc, core.source_id),
+                        &source_desc,
+                        &mut stream,
+                    )
+                    .await?;
+                }
+                Ok(msg) => {
+                    match msg {
+                        // This branch will be preferred.
+                        Either::Left(msg) => {
+                            match &msg {
+                                Message::Barrier(barrier) => {
+                                    if let Some(mutation) = barrier.mutation.as_deref() {
+                                        match mutation {
+                                            Mutation::Pause => stream.pause_stream(),
+                                            Mutation::Resume => stream.resume_stream(),
+                                            _ => (),
+                                        }
+                                    }
+
+                                    state_store_handler
+                                        .state_store
+                                        .commit(barrier.epoch)
+                                        .await?;
+
+                                    if let Some(vnode_bitmap) =
+                                        barrier.as_update_vnode_bitmap(self.actor_ctx.id)
+                                    {
+                                        // if _cache_may_stale, we must rebuild the stream to adjust vnode mappings
+                                        let (_prev_vnode_bitmap, _cache_may_stale) =
+                                            state_store_handler
+                                                .state_store
+                                                .update_vnode_bitmap(vnode_bitmap);
+                                    }
+
+                                    // Propagate the barrier.
+                                    yield msg;
```

Comment:
Let's only rebuild the stream when barrier comes (except the recovery process). The benefit is all changes are aligned to the barrier, and not to modify the stream at runtime.

Reply:
By doing so, the data rate is actually limited by the SPLIT_BATCH_SIZE and may be lower than the used-to-be best-effort strategy. Is this an expected feature?

Reply:
lets say `SPLIT_BATCH_SIZE=1024`, the reader stream keeps reading files for N epochs, and ends between barrier N and barrier N+1. There isn't a huge efficiency loss when idling for less than one epoch.
I believe this is going to make it easier managing a stream. Adding new files to the task queue is a kind of config change and we don't want let it happen often, just like what we do in the traditional source exec